airflow/contrib/hooks/opsgenie_alert_hook.py

```diff
@@ -0,0 +1,83 @@
+# or more contributor license agreements.  See the NOTICE file
+# distributed with this work for additional information
+# regarding copyright ownership.  The ASF licenses this file
+# to you under the Apache License, Version 2.0 (the
+# "License"); you may not use this file except in compliance
+# with the License.  You may obtain a copy of the License at
+#
+#   http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing,
+# software distributed under the License is distributed on an
+# "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+# KIND, either express or implied.  See the License for the
+# specific language governing permissions and limitations
+# under the License.
+#
+
+import json
+
+from airflow.hooks.http_hook import HttpHook
+from airflow.exceptions import AirflowException
+
+
+class OpsgenieAlertHook(HttpHook):
+    """
+    This hook allows you to post alerts to Opsgenie.
+    Accepts a connection that has an Opsgenie API key as the connection's password.
+    Each Opsgenie API key can be pre-configured to a team integration.
+    You can override these defaults in this hook.
+
+    :param http_conn_id: Http connection ID with host as "https://api.opsgenie.com/"
+                      and Opsgenie API key as the connection's password
+                      (e.g. "eb243592-faa2-4ba2-a551q-1afdf565c889")
+    :type http_conn_id: str
+    :param payload: Opsgenie API Create Alert payload values
+                    See https://docs.opsgenie.com/docs/alert-api#section-create-alert
+    :type payload: dict
+    :param proxy: Proxy to use to make the Opsgenie Alert API call
+    :type proxy: str
+    """
+    def __init__(self,
+                 http_conn_id=None,
+                 payload={},
+                 proxy=None,
+                 *args,
+                 **kwargs
+                 ):
+        super(OpsgenieAlertHook, self).__init__(*args, **kwargs)
+        self.http_conn_id = http_conn_id
+        self.api_key = None
+        self.payload = payload
+        self.proxy = proxy
+
+    def _get_api_key(self, http_conn_id):
+        """
+        Given a conn_id, return the api_key to use
+        :param http_conn_id: The conn_id provided
+        :type http_conn_id: str
+        :return: api_key (str) to use
+        """
+        if http_conn_id:
+            conn = self.get_connection(http_conn_id)
+            return conn.password
+        else:
+            raise AirflowException('Cannot get api_key: No valid conn_id '
+                                   'supplied')
+
+    def execute(self):
+        """
+        Remote Popen (actually execute the Opsgenie Alert call)
+        """
+        proxies = {}
+        if self.proxy:
+            # we only need https proxy for Opsgenie, as the endpoint is https
+            proxies = {'https': self.proxy}
+        self.api_key = self.api_key or self._get_api_key(self.http_conn_id)
```

Comment:
According to reivew in L53, maybe we could change, WDYT?
```suggestion
        api_key = self._get_api_key(self.http_conn_id)
```
also change variable name in L82

Reply:
@zhongjiajie My only worry was that means every time the `execute` is called it's doing another `get_connection` to get the `api_key`. What I was attempting to do was basically store the api_key on the object level for the hook, so that further calls will have the key already (basically keeping the `get_connection` to once call per instantiated hook.

If that is not a concern, I have no problem changing to your suggestion.

Reply:
I don't think `get_connection` out of constructor will call many time in scheduler? WDYT cc@mik-laj . Only in constructor mean `__init__` function will call many time, I agree

Reply:
In this case, I wasn't necessarily thinking of during the constructor but rather per `execute`.

For example, a use case is instantiating a single hook object and then calling `execute` more than once. The first call to `execute` will do the `get_connection`, subsequent calls will not.

Reply:
> In this case, I wasn't necessarily thinking of during the constructor but rather per `execute`.
> 
> For example, a use case is instantiating a single hook object and then calling `execute` more than once. The first call to `execute` will do the `get_connection`, subsequent calls will not.

I not sure about that.

Reply:
Each operator is performed in isolation. The previous state of the operation is not stored. It is possible to optimize the number of database queries during one execution. Each operator execution is a new Python interpreter. Airflow is designed for distributed systems where shared memory does not exist. Airflow is intended for distributed systems. It does not have shared memory. 

Reference: 
> In previous versions of the scheduler, user-supplied DAG definitions were parsed and loaded in the same process as the scheduler. Unfortunately, this made it possible for bad user code to adversely affect the scheduler process. For example, if a user DAG definition includes a `system.exit(-1)`, parsing the DAG definition would cause the scheduler process to exit.

https://cwiki.apache.org/confluence/display/AIRFLOW/Scheduler+Basics

https://github.com/apache/airflow/blob/master/airflow/executors/sequential_executor.py#L48
https://github.com/apache/airflow/blob/master/airflow/executors/local_executor.py#L86
https://github.com/apache/airflow/blob/master/airflow/executors/celery_executor.py#L66

Reply:
@mik-laj So it mean that
> For example, a use case is instantiating a single hook object and then calling execute more than once. The first call to execute will do the get_connection, subsequent calls will not.

will not happend, right?

Reply:
It makes sense, but in that case the logic is placed in the wrong place. We should use `cached_property`. This is better to explain the expected behavior.

Reply:
In order to facilitate this PR, and also since I wasn't sure if you meant `cached_property` as in the package, which would introduce another dependency, I just removed this 'caching' functionality entirely.

It is very possible that this may never happen anyways, since the majority of the use cases for this hook will most likely be through the operator, and even the small subset of users using the hook may never call execute more than once.